_posts/2025-05-09-opentelemetry-zero-code-instrumentation.md

```diff
@@ -0,0 +1,537 @@
+    environment:
+      OTEL_SERVICE_NAME: order.${NAMESPACE:-openfaas-fn}
+      OTEL_TRACES_EXPORTER: ${EXPORTER:-otlp}
+      OTEL_METRICS_EXPORTER: none
+      OTEL_LOGS_EXPORTER: none
+      OTEL_EXPORTER_OTLP_ENDPOINT: ${OTEL_EXPORTER_OTLP_ENDPOINT:-alloy.monitoring.svc.cluster.local:4317}
+      OTEL_EXPORTER_OTLP_TRACES_INSECURE: true
+```
+
+Build and deploy the function:
+
+```sh
+faas-cli up --filter order
+```
+
+Invoke the function and inspect the trace in Grafana. You should see the custom `process_order` span as part of the function invocation.
+
+```sh
+curl -i "http://127.0.0.1:8080/function/order?id=1"
+```
+
+![Screenshot of the Grafana Explorer showing an invocation trace with a custom span](/images/2025-05-opentelemetry/manual-span-trace.png)
+> Trace detail showing the function invocation with the custom span. The span has the order id attribute and events that we added to it.
+
+## OpenTelemetry on OpenFaaS Edge
+
+Collecting telemetry is also supported on [OpenFaaS Edge](https://docs.openfaas.com/edge/overview/). The Alloy collector can be deployed as an additional service. Detailed instructions on how to configure Alloy and add it to the `docker-compose.yaml` file are available in [our documentation](https://docs.openfaas.com/edge/open-telemetry/).
+
+Any instrumented function can be deployed to OpenFaaS Edge. Just make sure the `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable is set correctly.
+Modify the `stack.yaml` file and change the environment variable to `alloy:4317`. Or use the environment substitution to override it. In OpenFaaS Edge the service name can be used as the hostname.
+
+```sh
+# Use the stack.yaml from the previous section and override the OTLP endpoint.
+OTEL_EXPORTER_OTLP_ENDPOINT=alloy:4317 faas-cli up
+```
+
+## Conclusion
+
+In this tutorial we showed how OpenTelemetry zero-code instrumentation can be used to quickly get telemetry and traces for OpenFaaS functions without having to manually instrument your functions code.
+
+We ran through the process of deploying and configuring a [Collector](https://opentelemetry.io/docs/collector/), and tracing backend to collect and store traces for OpenFaaS functions.
+
+To support auto instrumentation some minor customisations to the OpenFaaS function templates are required. We walked through the steps to modify the Node.js and Python templates. With python we took it one step further and extended the auto instrumented traces with some custom spans.
+
+If you are using OpenFaaS on Kubernetes there is the option to use the [OpenTelemetry Operator for Kubernetes ](https://opentelemetry.io/docs/platforms/kubernetes/operator/) to [inject zero-code instrumentation](https://opentelemetry.io/docs/platforms/kubernetes/operator/automatic/). The Operator supports injecting and configuring auto-instrumentation libraries for .NET, Java, Node.js, Python and Go functions. It requires no changes to your existing functions, however it comes with a couple of tradeoffs.
```

Comment:
This note contains a lot of detail - I'd suggest a sub-heading, after you've said everything you want to say in the conclusion.